neural_compressor/onnxrt/algorithms/smooth_quant/smooth_quant.py

```diff
@@ -0,0 +1,653 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""SmoothQuant for onnxrt."""
+
+import copy
+import logging
+import os
+
+import numpy as np
+import onnx
+from onnx import helper, numpy_helper
+from onnx import onnx_pb as onnx_proto
+
+from neural_compressor.onnxrt.algorithms.smooth_quant.calibrator import Calibrator
+from neural_compressor.onnxrt.utils.onnx_model import ONNXModel
+from neural_compressor.onnxrt.utils.utility import (
+    _get_qrange_for_qType,
+    is_B_transposed,
+    quantize_data,
+    simple_progress_bar,
+)
+
+logger = logging.getLogger("neural_compressor")
+
+dtype_map = {
+    np.dtype("float32"): 1,
+    np.dtype("uint8"): 2,
+    np.dtype("int8"): 3,
+    np.dtype("int32"): 6,
+    np.dtype("int64"): 7,
+    np.dtype("float16"): 10,
+    np.dtype("double"): 11,
+}
+
+
+def get_quant_dequant_output(model, input_data, output_data, providers):
+    """Get loss between fp32 output and QDQ output.
+
+    Args:
+        model (object): model
+        input_data (numpy.ndarray): fp32 input
+        output_data (numpy.ndarray): fp32 output
+        providers (list): execution provider
+    """
+    import onnxruntime as ort
+
+    input_data = quant_dequant_data(input_data, 2, "asym")
+    sess = ort.InferenceSession(model.SerializeToString(), providers=providers)
+    preds = sess.run(None, {model.graph.input[0].name: input_data})
+    loss = np.sum(np.abs(output_data - preds) ** 2)
+    return loss
+
+
+def make_sub_graph(node, inits, input_data, output_data, opset, ir_version):
+    """Build a model with the specific node.
+
+    Args:
+        node (object): node
+        inits (list): initializer inputs of this node
+        input_data (numpy.ndarray): fp32 input
+        output_data (numpy.ndarray): fp32 output
+        opset (object): opset of the model
+        ir_version (object): ir_version of the model
+    """
+    from onnx import helper
+
+    input = helper.make_tensor_value_info(node.input[0], dtype_map[input_data.dtype], input_data.shape)
+    output = helper.make_tensor_value_info(node.output[0], dtype_map[output_data.dtype], output_data.shape)
+    graph = helper.make_graph([node], "sub_graph", [input], [output], inits)
+    model = helper.make_model(graph, opset_imports=opset)
+    model.ir_version = ir_version
+    return model
+
+
+def quant_dequant_data(data, qType=3, scheme="sym"):
+    """Quantize and then dequantize data.
+
+    Args:
+        data (numpy.ndarray): target data
+        qType (int): data type
+        scheme (str): sym or asym quantization
+    """
+    rmin, rmax, zero_point, scale, quantized_data = quantize_data(
+        data.flatten().tolist(), _get_qrange_for_qType(qType, False), qType, scheme
+    )
+    return ((quantized_data - zero_point) * scale).astype(data.dtype).reshape(data.shape)
+
+
+class ORTSmoothQuant:

```

Comment:
this algo interface should be common, just implementation different cross backends. it's better to align with other framework owners.

Reply:
do you mean extract a SmoothQuant base class?

Reply:
If sq needs to align, woq seems need align as well.